File: 409/main.cpp
/* 409. 最长回文串
 * 
 * 20200907
 * huao
 */

#include <string.h>

int longestPalindrome(char * s){
    int count[128];
    memset(count, 0, sizeof(int) * 128);

    for (int i = 0; s[i] != 0; i++) {
        count[s[i]]++;
    }
    int flag = 0;
    int length = 0;
    for (int i = 'A'; i <= 'z'; i++) {
        if (count[i] % 2 == 1) {
            flag = 1;
            length += count[i] - 1;
        } else {
            length += count[i];
        }
    }
    return length + flag;
}

File: 409/main.py
# 409. 最长回文串
#
# 20200907
# huao

class Solution:
    def longestPalindrome(self, s: str) -> int:
        count = [0 for i in range(52)]
        for c in s:
            if c.islower():
                count[ord(c)-97] += 1
            elif c.isupper():
                count[ord(c)-97+26] += 1
        flag = 0
        length = 0
        for i in count:
            if i % 2 == 1:
                flag = 1
                length += (i-1)
            else:
                length += i
        return length + flag


print(Solution().longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth"))
